### DISCUSSION TITLE
cannot start docker container v1.25.0

### DISCUSSION THREAD

**Discussion by bassarf** [dani-garcia/vaultwarden#2510] [Q&A]
### Subject of the issue

When I try to start my docker container it does not start.

### Deployment environment

* vaultwarden version: 1.25.0

* Install method: Docker

* Other relevant details:

### Steps to reproduce

Starting the container via:
```
docker run -d --name vaultwarden \
  -e ROCKET_TLS='{certs="/ssl/live/<domain>/fullchain.pem",key="/ssl/live/<domain>/privkey.pem"}' \
  -e SIGNUPS_ALLOWED=false \
  -e ADMIN_TOKEN=<token string> \
  -v /etc/letsencrypt/:/ssl/ \
  -v /home/<user>/programme/vaultwarden/vw-data/:/data/ \
  -p <port>:80 \
  --restart=unless-stopped \
  vaultwarden/server:latest
```

### Expected behaviour

container should start

### Actual behaviour

container does not start correctly. When typing `docker logs vaultwarden` I get the following error
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

when starting with an additional `RUST_BACKTRACE=1` I get the following
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: tokio::time::driver::Driver<P>::new
   4: tokio::runtime::driver::Driver::new
   5: tokio::runtime::builder::Builder::build
   6: rocket::async_main
   7: vaultwarden::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: tokio::time::driver::Driver<P>::new
   4: tokio::runtime::driver::Driver::new
   5: tokio::runtime::builder::Builder::build
   6: rocket::async_main
   7: vaultwarden::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by BlackDex** [comment]
Looks like the container isn't allowed to get the current time or something. Could you try to use `RUST_BACKTRACE=full` together with `LOG_LEVEL=debug`?

What is your Host OS?

**Comment by oven8Mitts** [comment]
Also getting this issue on:
Raspberry Pi 4 Model B Rev 1.1, running Raspbian Buster
ARM v7

Vaultwarden info:
sha256:f3ebede27f1cf5e78373c3c4a429cf1fdd8d6b13528a2b9ca4fb3cb7cc681ba9
 vaultwarden/server:latest


Here's my complete trace with debug on:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
stack backtrace:
   0:   0xb28d78 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::had5730c67ab50e4d
   1:   0x607090 - core::fmt::write::ha42b2480a34f3eb5
   2:   0xb21fe4 - std::io::Write::write_fmt::hfa663667fab44c82
   3:   0xb2ac60 - std::panicking::default_hook::{{closure}}::h6e8d770966f585fd
   4:   0xb2a8f0 - std::panicking::default_hook::h71b5cfb8a06cd56f
   5:   0xb2b1dc - std::panicking::rust_panic_with_hook::hfc13f45907194734
   6:   0xb2b0dc - std::panicking::begin_panic_handler::{{closure}}::h8bd28c15b24c2fb7
   7:   0xb29390 - std::sys_common::backtrace::__rust_end_short_backtrace::hef7cb296473d0b87
   8:   0xb2ae6c - rust_begin_unwind
   9:   0x5812f4 - core::panicking::panic_fmt::hc45184285ee465d2
  10:   0x5813c8 - core::result::unwrap_failed::h51f4224fc12cbf0e
  11:   0xb70a88 - tokio::time::driver::Driver<P>::new::hfb66ef6768de0218
  12:   0xb5f658 - tokio::runtime::driver::Driver::new::hb1ee45257e549b73
  13:   0xb5f124 - tokio::runtime::builder::Builder::build::hed4ce3523a05a186
  14:   0xe94f0c - rocket::async_main::h077ba2654118e619
  15:   0xde5fbc - vaultwarden::main::h433acfc7f8dab814
  16:   0xe0259c - std::sys_common::backtrace::__rust_begin_short_backtrace::heeb0eb3b70d82df1
  17:  0x1198670 - std::rt::lang_start::{{closure}}::ha2f1a90a8484b316
  18:   0xb1b170 - std::rt::lang_start_internal::ha117b08a7dcebcaf
  19:   0xde5ff0 - main
  20: 0xb6b69a20 - __libc_start_main
```

**Comment by moduraziel** [comment]
For what it's worth, I am redeploying vaultwarden after trying it out a couple days ago (and it working) and I am having the exact same errors. Something wrong with the image maybe?

**Comment by bassarf** [comment]
> Looks like the container isn't allowed to get the current time or something. Could you try to use `RUST_BACKTRACE=full` together with `LOG_LEVEL=debug`?
> 
> What is your Host OS?

Thank you for your fast reply. Here is the log output with `RUST_BACKTRACE=full` and `LOG_LEVEL=debug` (Its basically the same as from @99ytrewq991)

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
stack backtrace:
   0:   0xb58d78 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::had5730c67ab50e4d
   1:   0x637090 - core::fmt::write::ha42b2480a34f3eb5
   2:   0xb51fe4 - std::io::Write::write_fmt::hfa663667fab44c82
   3:   0xb5ac60 - std::panicking::default_hook::{{closure}}::h6e8d770966f585fd
   4:   0xb5a8f0 - std::panicking::default_hook::h71b5cfb8a06cd56f
   5:   0xb5b1dc - std::panicking::rust_panic_with_hook::hfc13f45907194734
   6:   0xb5b0dc - std::panicking::begin_panic_handler::{{closure}}::h8bd28c15b24c2fb7
   7:   0xb59390 - std::sys_common::backtrace::__rust_end_short_backtrace::hef7cb296473d0b87
   8:   0xb5ae6c - rust_begin_unwind
   9:   0x5b12f4 - core::panicking::panic_fmt::hc45184285ee465d2
  10:   0x5b13c8 - core::result::unwrap_failed::h51f4224fc12cbf0e
  11:   0xba0a88 - tokio::time::driver::Driver<P>::new::hfb66ef6768de0218
  12:   0xb8f658 - tokio::runtime::driver::Driver::new::hb1ee45257e549b73
  13:   0xb8f124 - tokio::runtime::builder::Builder::build::hed4ce3523a05a186
  14:   0xec4f0c - rocket::async_main::h077ba2654118e619
  15:   0xe15fbc - vaultwarden::main::h433acfc7f8dab814
  16:   0xe3259c - std::sys_common::backtrace::__rust_begin_short_backtrace::heeb0eb3b70d82df1
  17:  0x11c8670 - std::rt::lang_start::{{closure}}::ha2f1a90a8484b316
  18:   0xb4b170 - std::rt::lang_start_internal::ha117b08a7dcebcaf
  19:   0xe15ff0 - main
  20: 0xb6ac6a20 - __libc_start_main
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62
stack backtrace:
   0:   0xaf8d78 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::had5730c67ab50e4d
   1:   0x5d7090 - core::fmt::write::ha42b2480a34f3eb5
   2:   0xaf1fe4 - std::io::Write::write_fmt::hfa663667fab44c82
   3:   0xafac60 - std::panicking::default_hook::{{closure}}::h6e8d770966f585fd
   4:   0xafa8f0 - std::panicking::default_hook::h71b5cfb8a06cd56f
   5:   0xafb1dc - std::panicking::rust_panic_with_hook::hfc13f45907194734
   6:   0xafb0dc - std::panicking::begin_panic_handler::{{closure}}::h8bd28c15b24c2fb7
   7:   0xaf9390 - std::sys_common::backtrace::__rust_end_short_backtrace::hef7cb296473d0b87
   8:   0xafae6c - rust_begin_unwind
   9:   0x5512f4 - core::panicking::panic_fmt::hc45184285ee465d2
  10:   0x5513c8 - core::result::unwrap_failed::h51f4224fc12cbf0e
  11:   0xb40a88 - tokio::time::driver::Driver<P>::new::hfb66ef6768de0218
  12:   0xb2f658 - tokio::runtime::driver::Driver::new::hb1ee45257e549b73
  13:   0xb2f124 - tokio::runtime::builder::Builder::build::hed4ce3523a05a186
  14:   0xe64f0c - rocket::async_main::h077ba2654118e619
  15:   0xdb5fbc - vaultwarden::main::h433acfc7f8dab814
  16:   0xdd259c - std::sys_common::backtrace::__rust_begin_short_backtrace::heeb0eb3b70d82df1
  17:  0x1168670 - std::rt::lang_start::{{closure}}::ha2f1a90a8484b316
  18:   0xaeb170 - std::rt::lang_start_internal::ha117b08a7dcebcaf
  19:   0xdb5ff0 - main
  20: 0xb6b1aa20 - __libc_start_main
```
(I do not know why `docker logs vaultwarden` gives me the same stacktrace twice)

I also run it on a Raspberry PI 4 with buster

**Comment by arctic-alpaca** [comment]
This might be related to this issue: https://github.com/moby/moby/issues/40734

If it is, running the container as privileged or seccomp=unconfined should work as a quick and dirty work-around. More permanent solutions are discussed in the linked issue and the other issues linked there.

**Comment by BlackDex** [comment]
Strange, that part of the code is linked to netbsd. Not sure from the top of my head if that is the _default_ Linux for Rust, but that would be strange.

**Comment by BlackDex** [comment]
Maybe a strange thing, but could you both try the Alpine image maybe?

**Comment by bassarf** [comment]
> This might be related to this issue: [moby/moby#40734](https://github.com/moby/moby/issues/40734)
> 
> If it is, running the container as privileged or seccomp=unconfined should work as a quick and dirty work-around. More permanent solutions are discussed in the linked issue and the other issues linked there.

Thank you for your reply. Indeed using the flag `--security-opt seccomp=unconfined` works. The server starts and the client is able to sync.



> Maybe a strange thing, but could you both try the Alpine image maybe?

Thank you for your reply... this also works.. even without the `seccomp` flag above.

**Comment by moduraziel** [comment]
Im still learning my way around Docker/Portainer...how do I add this flag to my container? Is it just an ENV variable? And how do I access the Alpine image?

**Comment by bassarf** [comment]
> Im still learning my way around Docker/Portainer...how do I add this flag to my container? Is it just an ENV variable? And how do I access the Alpine image?

in order to use the alpine image you just use `vaultwarden/server:alpine` instead of `vaultwarden/server:latest`.
in order to use the flag you just add `--security-opt seccomp=unconfined` to your docker run command.

**Reply by bitbull06** [reply]
> > Im still learning my way around Docker/Portainer...how do I add this flag to my container? Is it just an ENV variable? And how do I access the Alpine image?
> 
> in order to use the alpine image you just use `vaultwarden/server:alpine` instead of `vaultwarden/server:latest`. in order to use the flag you just add `--security-opt seccomp=unconfined` to your docker run command.

The alpine image instead of the latest does the trick for me too, even without the seccomp flag.
However, any idea how to set the seccomp flag in Portainer? I'like to switch back to the latest tag - what's even the difference?

**Reply by BlackDex** [reply]
using either `:alpine` or `:latest` will result in the exact same version. The difference is the `:alpine` Runs on an Alpine based image, the other is based upon Debian.

**Reply by bitbull06** [reply]
okay, thanks. look indeed like exactly the same version. but what about the `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:353:62` error being thrown with latest (Debian) image but not present in Alpine? this prevents the container from being started with Debian image.
And regarding my question about the flag?

**Reply by BlackDex** [reply]
I have no clue regarding the flag, as far as i know podman is rootless and maybe it is not possible with podman.

That error happens on Debian because it used glibc and it doesn't happen on Alpine becuse it uses MUSL, they both use different ways of handling those calls.

**Reply by moduraziel** [reply]
to start vaultwarden with the --security-opt seccomp=unconfined flag, you need to run it from ssh. As far as I can tell, there is no way to add the flag in portainer.

`docker run --security-opt seccomp=unconfined vaultwarden/server:latest`

**Comment by moduraziel** [comment]
Thanks for the quick reply. Is it possible to use the seccomp flag through portainer or do I need to run the container through CLI to add the command? I was able to enable Privileged Mode in the GUI for the container, which seemed to work, I'm just curious how you would use that flag. Alpine appears to be working.

Also, from a security perspective, which would be the better choice: Alpine, run flag, or privileged mode?

**Comment by oven8Mitts** [comment]
Sadly, the Alpine image did not fix the issue on my end, although unconfining seccomp did resolve the issue for me.

**Comment by BlackDex** [comment]
It could be that the docker host has an outdated OS in the sens that it doesn't understand the correct calls. This seems to be a docker<>kernel issue and would probably cause all docker images which are using a newer OS based image and does the same time call break those images. Now I'm not sure we can really solve that.

The reason it is breaking with the newer version is because the Vaultwarden images are using an updated base image since this release compared to the previous one.

reverting back to an older release of the base images isn't really something i would want to do, since those images are using older library versions.

**Comment by moduraziel** [comment]
I am running Buster instead of Bullseye due to another app's requirements. Could that be an explanation for these issues?

**Comment by BlackDex** [comment]
From what i understand is that since buster is using an older kernel than bullseye, it doesn't support some system-calls out-of-the box. And that is why you need to use `--security-opt seccomp=unconfined` to allow these calls to work.

**Comment by moduraziel** [comment]
Ah that makes a lot of sense, then. And that would explain why I got it working a few days ago but not more recently, until I added that flag

**Comment by bassarf** [comment]
It looks like this is a "buster only" problem, since everyone with this issue is running buster (as far as I can see it).
Thus this seems like a "backwards compatibility" issue and I guess won't be fixed (if I understood BlackDex's comment correctly). 
So I guess this issue can be closed?!

Maybe a note in the wiki, readme.md and/or release notes, that one might have problems when running buster would be good.
The pihole project had a similar issue with seccomp. Their solution was (as far as I can tell) to not support buster anymore:
https://github.com/pi-hole/docker-pi-hole/releases/tag/2022.04.2

**Comment by bassarf** [comment]
Just upgraded to bullseye and everything works fine now with the `latest` image

**Comment by Kamuno** [comment]
Just wanted to say that I had this issue come up when upgrading from 1.24.0 to 1.25.0 on my raspberry pi 3.
Switching to the alpine tag fixed the issue for me.

**Comment by GeekCornerGH** [comment]
Got no issues on testing branch
